language-plutus-napkin/src/Language/PlutusCore/PrettyPrint.hs

```diff
@@ -0,0 +1,22 @@
+{-# LANGUAGE FlexibleContexts #-}
+
+module Language.PlutusCore.PrettyPrint ( rewriteTerm
+                                       ) where
+
+import           Data.Functor.Foldable
+import qualified Data.IntMap                    as IM
+import           Language.PlutusCore.Identifier
+import           Language.PlutusCore.Parser
+import           Language.PlutusCore.Type
+import           PlutusPrelude
+
+-- | A monadic catamorphism
+cataM :: (Recursive t, Traversable (Base t), Monad m) => (Base t a -> m a) -> (t -> m a)
+cataM phi = c where c = phi <=< traverse c . project
+
+rewriteTerm :: (IdentifierState, Program a) -> Either ParseError (Program a)
```

Comment:
What is the term rewriting about? It would be helpful if you could add comments to all non-trivial functions.

I think, I see what your intention is. You are rewriting program terms to use `PrintVar` terms instead of `Var` terms to prepare them for pretty printing. I don't think, we should do that. It clutters the abstract syntax with the `PrintVar` constructor. Now, whenever we pass a value of type `Term a`, we have to add an *informal* side condition along the line of "this argument is not supposed to include subterms with `PrintVar`" and then in the code what do we do if we come across one of those, where we are not supposed to? Raise an `error`? That will lead to some function being unnecessarily partial, which is often a source of unexpected bugs.